R/weight.R

```diff
@@ -41,33 +43,216 @@ weight <- function (x) {
+    if (is.variable(x)) {
+        ds <- loadDataset(datasetReference(self(x)))
+        return(identical(weight(ds), x))
+    } else {
+        return(FALSE)
+    }
+}
 
 #' @rdname weight
 #' @export
-setMethod("weightVariables", "VariableCatalog", function (x) {
+setMethod("is.weight<-", "NumericVariable", function(x, value) {
+    if (!(is.logical(value)  && length(value) == 1)) {
+        halt("Right hand side must be TRUE or FALSE")
+    }
+    ds <- loadDataset(datasetReference(self(x)))
+    varcat <- allVariables(ds)
+    if (value) {
+        weight(ds) <- x
+    } else {
+        weight(ds) <- NULL
+    }
+    x <- refresh(x)
+    return(x)
+})
+
+#' Change which variables can be set as weights
+#'
+#' Editors can change which variables can be set as the weighting variable
+#' for a dataset. For instance if several weights have been calculated they
+#' can let the user choose which of those variables to use a weight, but prevent
+#' the user from choosing other variables as weight. This function allows you
+#' to get and set the `weightVariables` of a dataset, which are the variables which
+#' can be set as the dataset [weight()].
+#'
+#' @param x a dataset or variable catalog
+#' @param value For the setter variables to set as the `weightVariables` this can
+#' be a numeric Crunch variable, list of numeric Crunch variables or
+#' a character vector with the aliases of numeric Crunch variables.
+#'
+#' @return For the getter the variables which can be set as the dataset's weight. For
+#' the setter the dataset or variable catalog duly modified.
+#' @seealso [weight()] [makeWeight()]
+#' @name weightVariables
+#' @examples
+#' \dontrun{
+#' weightVariables(ds) <- ds$weight
+#' weightVariables(ds) <- list(ds$weight, ds$weight2)
+#' weightVariables(ds) <- NULL
+#' weightVariables(ds) <- c("weight", "weight2")
+#' #To append a weight to existing weights
+#' weightVariables(ds) <- c(weightVariables(ds), "weight3")
+#' #To remove a weight
+#' weightVariables(ds) <- setdiff(weightVariables(ds), "weight3")
+#' }
+#' @name weightVaraibles
+#' @aliases is.weightVariable<- weightVariables<-
+#' 
+#' @export
+NULL
+
+#' @rdname weightVariables
+#' @export
+setMethod("weightVariables", "CrunchDataset",
+    function(x) weightVariables(allVariables(x)))
+
+#' @rdname weightVariables
+#' @export
+setMethod("weightVariables", "VariableCatalog", function(x) {
     ## Get weight variable order
     ord <- VariableOrder(crGET(shojiURL(x, "orders", "weights")))
     ## Subset weight variable catalog with it
     vars <- x[ord]
     ## Return the names/aliases
     if (length(vars)) {
         return(sort(vapply(index(vars), vget(namekey(x)), character(1),
-            USE.NAMES=FALSE)))
+            USE.NAMES = FALSE)))
     } else {
         return(c())
     }
 })
 
+#' @rdname weightVariables
+#' @export
+setMethod("weightVariables<-", "CrunchDataset", function(x, value) {
+    weightVariables(allVariables(x)) <- value
+    x <- refresh(x)
```

Comment:
This `refresh` shouldn't be necessary; the setting on the one above should do the necessary cache busting.